plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinHighLevelFunctionParameterInfoHandler.kt

```diff
@@ -0,0 +1,451 @@
+                }
+
+                // Find candidate matching the one in CandidateInfo
+                // TODO: restoreSymbol does not work for static members (see Nullability.kt test)
+                // and functional values (see FunctionalValue*.kt tests)
+                val candidateToMatch = candidateInfo.candidate.restoreSymbol() ?: return@analyse
+                val candidates = resolvedCall?.targetFunction?.candidates ?: return@analyse
+                val candidate = candidates.firstOrNull { areEqual(it, candidateToMatch) } ?: return@analyse
+
+                // For array set calls, we only want the index arguments in brackets, which are all except the last (the value to set).
+                val isArraySetCall = candidate.callableIdIfNonLocal?.let {
+                    val isSet = it.callableName == OperatorNameConventions.SET
+                    isSet && callElement is KtArrayAccessExpression
+                } ?: false
+                // TODO: Get substituted value parameters. Not currently available in HL API. See SubstituteFromArguments*.kt tests
+                val valueParameters = candidate.valueParameters.let { if (isArraySetCall) it.dropLast(1) else it }
+
+                // TODO: When resolvedCall is KtFunctionalTypeVariableCall, the candidate is FunctionN.invoke() and parameter names are "p1", "p2", etc.
+                // We need to get the type of the target variable, and retrieve the parameter names from the type (KtFunctionalType).
+                // The names need to be added to KtFunctionalType (currently only types are there) and populated in KtSymbolByFirBuilder.TypeBuilder.
+
+                val parameterToText = valueParameters.associateWith { parameter ->
+                    // TODO: Add hasSynthesizedParameterNames to HL API.
+                    // See resolveValueParameters() in core/descriptors.jvm/src/org/jetbrains/kotlin/load/java/lazy/descriptors/LazyJavaScope.kt
+                    val hasSynthesizedParameterNames = false
+                    renderParameter(parameter, includeName = !hasSynthesizedParameterNames)
+                }
+
+                // TODO: The argument mapping should also be per-candidate once we have all candidates available.
+                val argumentMapping = resolvedCall.argumentMapping
+
+                // TODO: This should be changed when there are multiple candidates available; need to know which one the call is resolved to
+                val isCallResolvedToCandidate = candidates.size == 1
+
+                candidateInfo.callInfo = CallInfo(
+                    callElement,
+                    valueArguments,
+                    arguments,
+                    argumentMapping,
+                    valueParameters,
+                    parameterToText,
+                    isCallResolvedToCandidate
+                )
+            }
+        }
+    }
+
+    private fun areEqual(first: KtFunctionLikeSymbol, second: KtFunctionLikeSymbol): Boolean {
```

Comment:
That would be better. Are you comparing the signatures then? That would be safer. What I'm doing works because there shouldn't be any changes in the PSI outside of the function call itself being updated.

Reply:
There is not need to compare symbols by the signature as `KtSymbol` may be used only inside the read action it was created in, so it should never be changed during it lifetime

Reply:
@darthorimar Not sure if I understood your comment correctly. The two symbols compared here are going to be from different read actions: one from `findElementForParameterInfo` and the other from `updateParameterInfo` (restored). So these may be two different symbols.

Reply:
Ah across different session would be tricky...

Actually, would it be possible to populate the `CallInfo` directly when creating `CandidateInfo`?

Reply:
At first, I thought I could populate all the `CallInfo` in `findElementForParameterInfo`. But it shouldn't be done there because `findElementForParameterInfo` is called once when the parameter info popup is requested, but `updateParameterInfo` is called once and then _also_ every time the user types something while the popup is open (the highlighting dynamically changes). So you still need to resolve again, because things like `argumentMapping` and `isCallResolvedToCandidate` can change as the user types.

Reply:
@punzki There should never be a case when you meet two symbols from different sessions in the same scope. Every `KtSymbol` should be used ONLY inside `analyse {..}` block and cannot be leaked outside it. So, there should not be a situation when two symbols from different `analyse {..}` blocks are available together in the same scope. So, there is no need to compare them

Reply:
About restoring symbols:
Consider the following code

```
val ktFunction: KtNamedFunction = ...

val ponter = readAction {
	analyse(context) { // analysis context #1
		val symbol: KtSymbol = ktFunction.getSymbol() // symbol A
		symbol.createPointer()
	}
}

// modify psi  somehow


readAction {
	analyse(context) {  // analysis context #2
		val symbol: KtSymbol = pointer.restore() // symbol B
		val otherSymbol: ktSymbol = ...
		otherSymbol == symbol
	}
}

```

The symbol `A` you create belongs to `analysis context #1`. You can restore that symbol by pointer only in some analysis context. So, when you restore your symbol by `pointer` you get the symbol `B`. Even symbols `A` and `B` represents the same declaration, those symbols are different. The symbol `A` belongs to  `analysis context #1`, meanwhile symbol `B` belongs to `analysis context #2` in which it was restored. So, them you compare `otherSymbol` with `symbol` inside `analysis context #2` they will both belong to the same  `analysis context #2`

Reply:
Using `==` works, but there seems to be an issue with symbol restoration for Java symbols:
```
val javaCandidateSymbol = ...
val pointer = javaCandidateSymbol.createPointer()
val restoredSymbol = pointer.restoreSymbol()
assert(javaCandidateSymbol == restoredSymbol) // BOOM
```
The root cause seems to be HL API creating new `ScopeSession()` every time it is getting unsubstituted scope during restoration, and therefore a new enhanced `FirFunction` is created and a new symbol is as well (instead of using the old enhanced `FirFunction` in the cache).

Changed to use `==`, and disabled the relevant test for now.